File: app/models/job.rb
# Copyright 2012 Twitter, Inc. and others.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

require 'thread'

# Represents a job that is executing in Resque.
#
# The background thread can periodically update the job's completion attribute
# to provide status updates on progress -- the job's percent completion should
# always be job.completed / job.total
#
# Attributes
#
# resque_job:   The UUID of the Resque job.
# title:        A title, displayed to Clockwork Raven users
# complete_url: The URL to direct the user to when to job has completed
# back_url:     The URL to direct the user to if they wish to go back to their
#               previous screen while the job executes
class Job < ActiveRecord::Base
  has_one :evaluation

  # Processes <param> using the ThreadPoolProcessor <worker_class>
  def run processor_class, items, options = {}
    options[:items] = items
    self.processor = processor_class
    self.resque_job = processor_class.create(options)
    self.save!
  end

  # Fetches this job's status from Redis
  def status_hash
    if self.resque_job
      return @status_hash if instance_variable_defined? :@status_hash
      @status_hash = Resque::Plugins::Status::Hash.get(self.resque_job)
    else
      nil
    end
  end

  # Returns the name of this job as a symbol (:new, :running, :done, :error, or
  # :killed)
  def status_name
    return :new unless resque_job
    return :done unless status_hash

    case status_hash.status
    when 'queued'
      :new
    when 'completed'
      :done
    when 'failed'
      :error
    when 'killed'
      :killed
    when 'working'
      :running
    else
      :done
    end
  end

  # Returns the percentage completion of this job (between 0 and 100),
  # based on this job's "completed" and "total" properties
  def percentage
    return 100 if status_name == :done
    return 0 if status_name == :new

    status_hash.pct_complete
  end

  # Returns the number of total steps in this job
  def total
    return status_hash.total if status_hash and status_hash.total
    return evaluation.tasks.count if evaluation
    return 1
  end

  # Returns the number of steps this job has compeleted
  def completed
    return total if status_name == :done
    return 0 if status_name == :new
    return status_hash.num || 0
  end

  # If this job's status is :error or :killed, returns a friendly error message
  # with suggestions for what action the user should take.
  #
  # If this job's status is not :error or :killed, returns nil.
  def error
    case status_name
    when :error
      status_hash.message + "\n\n" + processor::KILL_MESSAGE.strip_heredoc
    when :killed
      "Killed.\n\n" + processor::KILL_MESSAGE.strip_heredoc
    else
      nil
    end
  end

  # returns true iff the task has ended (status name is either :done, :failed
  # or :killed)
  def ended?
    return false unless resque_job
    return true unless status_hash

    !status_hash.killable?
  end

  # Asks for this job to be killed.
  def kill!
    return if ended?

    Resque::Plugins::Status::Hash.kill(self.resque_job)
  end

  # Override #processor and #processor= to store a string and load a
  # constant
  def processor
    (self[:processor] || 'Job::ThreadPoolProcessor').constantize
  end
  def processor= new_processor
    self[:processor] = new_processor.name
  end

  # subclass Job::ThreadPoolProcessor to run jobs.
  class ThreadPoolProcessor
    include Resque::Plugins::Status

    # Override this constant to change the name of the job.
    NAME = "Processing"

    # Override this constant to provide a friendly message to isplay to the user
    # if this job terminate in an error or is killed. Provide suggestion for
    # what the user should do.
    KILL_MESSAGE = "Job may have been partially completed"

    # override this to process and item.
    def process item
    end

    # override this to do stuff before processing starts
    def before
    end

    # override this to do stuff after processing ends
    def after
    end

    # use this to increment the completion counter as your job executes.
    def increment_completion incr=1
      @progress_lock.synchronize {
        @current += incr
        at(@current, @total, "At #{@current} of #{@total}")
      }
    end

    # don't override this
    def perform
      @options = options
      @items = options['items']
      @total = @items.length
      @current = 0
      @progress_lock = Mutex.new

      before

      increment_completion 0

      Threading.thread_pool @items do |item|
        process item
        increment_completion
      end

      after
    end
  end
end

File: app/assets/javascripts/job.js
/* Copyright 2012 Twitter, Inc. and others.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

// job.rb: Code to automatically update the job status page to show
//         progress of the page



$(function() {
    if($('body').attr('class') != "jobs show") {
        // we're not on the job status page; abort
        return;
    }

    Jobs.startUpdating();
})

Jobs = {
    // Time, in ms, between updates
    UPDATE_INTERVAL: 2000,
    UPDATE_URL: window.location.href + '.json',
    KILL_URL: window.location.href + '/kill.json',

    startUpdating: function() {
        Jobs.updateIntervalId = setInterval(Jobs._update, Jobs.UPDATE_INTERVAL);
    },

    stopUpdating: function() {
        if(Jobs.updateIntervalId) {
            clearInterval(Jobs.updateIntervalId);
        }
    },

    killJob: function() {
        if (confirm("Are you sure?")) {
        Jobs.stopUpdating();
            $.ajax({
                type: 'POST',
                url: Jobs.KILL_URL,
                success: function() {
                    Jobs.killButton().text('Killed.').attr('disabled', 'disabled');
                    Jobs.progressBar().addClass('progress-danger');
                    Jobs.progressBar().removeClass('active');
                },
                error: function() {
                    Jobs.killButton().text('Failed. Retry.').removeAttr('disabled');
                },
                complete: function() {
                    Jobs._update();
                    Jobs.startUpdating();
                }
            });
            Jobs.killButton().text('Killing...').attr('disabled', 'disabled');
        }
    },

    backButton: function() {
        return $('.btn-danger');
    },

    killButton: function() {
        return $('.btn-inverse');
    },

    continueButton: function() {
        return $('.btn-success');
    },

    progressBar: function() {
        return $('.progress');
    },

    progressText: function() {
        return $('.progress-text');
    },

    queuedText: function() {
        return $('.queued');
    },

    _update: function() {
        $.getJSON(Jobs.UPDATE_URL, Jobs._handleData);
    },

    _handleData: function(data) {
        // update progress bar and text
        $('#progress_bar').css('width', data.percentage + '%');
        $('.completed').text(data.completed);
        $('.total').text(data.total);

        // show error if there is one
        if(data.status_name == 'error' || (data.status_name == 'killed')) {
            $('.alert').slideDown();
        }
        else {
            $('.alert').slideUp();
        }

        $('.alert-content').text(data.error)

        // if the job isn't new, hide the "Queued" text and show progress
        if(data.status_name != 'new') {
            Jobs.progressText().show();
            Jobs.queuedText().hide();
        }

        // if the job has completed, un-disable the continue button
        if(data.status_name == 'done') {
            Jobs.continueButton().removeAttr('disabled');
            Jobs.progressBar().addClass('progress-success');
        }
        else {
            Jobs.continueButton().attr('disabled', 'disabled');
        }

        // red progress bar for killed or failed jobs
        if((data.status_name == 'error') || (data.status_name == 'killed')) {
            Jobs.progressBar().addClass('progress-danger');
        }

        // if the job has ended, style the progress bar, disable kill,
        // and stop updating this page.
        if(data['ended?'] == true) {
            Jobs.killButton().attr('disabled', 'disabled');
            Jobs.progressBar().removeClass('active');
            Jobs.stopUpdating();
        }
        else {
            Jobs.killButton().removeAttr('disabled');
        }
    }
}